### PR TITLE
Fixes #2452  Added two new text properties

### LINKED ISSUE #2452: [FEAT] Additional properties NLP
Add additional properties for the NLP sub package.

We have a lot of checks in the NLP sub package that rely on high quality text properties.
Currently we have some which can be seen at text_properties.py.

We need to add more, can be based on simple internal python methods like text length or complex that uses external models such as formality.

When suggesting a new property - include what is the run time for the property and what are the dependencies it has

### PR DESCRIPTION
**Pull request by hjain5164**
Fixes #2452  

#### What does this implement/fix? Explain your changes.
I have added two new text properties:
- Lexical Density: Percentage of unique words in the text, rounded up to 2 decimal digits.
- Noun count: Number of unique noun words in the text.

Time taken for the two metrics on the tweet dataset (in seconds): 
* Lexical Density: 06.189267
* Unique Noun Count: 22.837775

#### Any other comments?
Added the test to calculate these two new properties on the tweet test dataset and asserted with the first 10 results.

### CODE REVIEW

File: tests/nlp/utils/test_properties.py [hunk 1]
```diff
@@ -38,3 +38,17 @@ def test_calculate_default_properties():
 
     # Assert
     assert_that(result, equal_to({'Toxicity': [0] * 20_000}))
+
+
+def test_skipping_long_calculation_properties(tweet_emotion_train_test_textdata):
```
Comment (Nadav-Barak) [inline tests/nlp/utils/test_properties.py:43]: Add a test that verifies if those tests can handle different types of None in the text (behaviour should be to ignore them)

File: deepchecks/nlp/utils/text_properties.py [hunk 2]
```diff
@@ -204,7 +211,9 @@ def formality(raw_text: Sequence[str], device: Optional[int] = None) -> List[flo
     {'name': 'Subjectivity', 'method': subjectivity, 'output_type': 'numeric'},
     {'name': 'Toxicity', 'method': toxicity, 'output_type': 'numeric'},
     {'name': 'Fluency', 'method': fluency, 'output_type': 'numeric'},
-    {'name': 'Formality', 'method': formality, 'output_type': 'numeric'}
+    {'name': 'Formality', 'method': formality, 'output_type': 'numeric'},
+    {'name': 'Lexical Density', 'method': lexical_density, 'output_type': 'numeric'},
+    {'name': 'Noun Count', 'method': noun_count, 'output_type': 'numeric'},
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:216]: We discussed doing Unique Noun Count, I think it will be better :)

File: deepchecks/nlp/utils/text_properties.py [hunk 3]
```diff
@@ -193,6 +193,13 @@ def formality(raw_text: Sequence[str], device: Optional[int] = None) -> List[flo
     classifier = get_transformer_pipeline('formality', model_name, device=device)
     return [x['score'] if x['label'] == 'formal' else 1 - x['score'] for x in classifier(raw_text)]
 
+def lexical_density(raw_text: Sequence[str]) -> List[str]:
+    """Return list of floats of lexical density."""
+    return [round(len(set(textblob.TextBlob(text).words)) * 100 / len(textblob.TextBlob(text).words), 2) for text in raw_text]
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:198]: would split this into several line for better readability

File: deepchecks/nlp/utils/text_properties.py [hunk 4]
```diff
@@ -193,6 +193,13 @@ def formality(raw_text: Sequence[str], device: Optional[int] = None) -> List[flo
     classifier = get_transformer_pipeline('formality', model_name, device=device)
     return [x['score'] if x['label'] == 'formal' else 1 - x['score'] for x in classifier(raw_text)]
 
+def lexical_density(raw_text: Sequence[str]) -> List[str]:
+    """Return list of floats of lexical density."""
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:197]: add a few more word and also maybe link to website explaining what is lexical_density

**Review by Nadav-Barak — COMMENTED** [review]
Hey, looks good!
Few additional things:
1.  add the properties to doc file:  nlp_properties.rst
2. Provide statistics on the calculation run time of those properties for the tweet dataset

File: deepchecks/nlp/utils/text_properties.py (hunk 2 above)
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py:216]: Yeah, my bad, actually the logic is written to calculate the unique noun counts only, but the naming is just Noun Count. I will change the name to Unique Noun Count!!

File: deepchecks/nlp/utils/text_properties.py (hunk 3 above)
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done

File: deepchecks/nlp/utils/text_properties.py (hunk 4 above)
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done

File: tests/nlp/utils/test_properties.py (hunk 1 above)
Comment (hjain5164) [reply tests/nlp/utils/test_properties.py:42]: Done

File: deepchecks/nlp/utils/text_properties.py [hunk 5]
```diff
@@ -193,6 +193,38 @@ def formality(raw_text: Sequence[str], device: Optional[int] = None) -> List[flo
...
+            result.append(lexical_density)
+        except:
+            result.append(np.nan)
+    return result
+
+def unique_noun_count(raw_text: Sequence[str]) -> List[str]:
+    """Returns list of integers of number of unique noun words in the text"""
+    result = []
+    for text in raw_text:
+        try:
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:218]: try except is a time consuming operation in python, better to use a if condition:

if not pd.isna(text):

File: deepchecks/nlp/utils/text_properties.py [hunk 6]
```diff
@@ -193,6 +193,38 @@ def formality(raw_text: Sequence[str], device: Optional[int] = None) -> List[flo
...
+def lexical_density(raw_text: Sequence[str]) -> List[str]:
+    """
+    Lexical density is the percentage of unique words in a given text. For more 
+    information: https://en.wikipedia.org/wiki/Lexical_density
+
+    Returns a list of floats of lexical density.
+    """
+    result = []
+    for text in raw_text:
+        try:
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:205]: same here

File: deepchecks/nlp/utils/text_properties.py [hunk 7]
```diff
@@ -193,6 +193,38 @@ def formality(raw_text: Sequence[str], device: Optional[int] = None) -> List[flo
     classifier = get_transformer_pipeline('formality', model_name, device=device)
     return [x['score'] if x['label'] == 'formal' else 1 - x['score'] for x in classifier(raw_text)]
 
+def lexical_density(raw_text: Sequence[str]) -> List[str]:
+    """
+    Lexical density is the percentage of unique words in a given text. For more 
+    information: https://en.wikipedia.org/wiki/Lexical_density
+
+    Returns a list of floats of lexical density.
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:201]:
```suggestion
    Return a list of floats of lexical density per text sample.
    
    Lexical density is the percentage of unique words in a given text. For more 
    information: https://en.wikipedia.org/wiki/Lexical_density
```

Comment (Nadav-Barak) [reply deepchecks/nlp/utils/text_properties.py:201]: Other wise it will fail on the code style CI test

File: tests/nlp/utils/test_properties.py [hunk 8]
```diff
@@ -38,3 +37,20 @@ def test_calculate_default_properties():
...
+
+    # Act
+    result = calculate_default_properties(test_text, include_properties=['Lexical Density', 'Unique Noun Count'])[0]
+    result_none_text = calculate_default_properties([None], include_properties=['Lexical Density', 'Unique Noun Count'])[0]
+
+    # Assert
+    assert_that(result['Lexical Density'][0: 10], equal_to([94.44, 93.75, 100.0, 91.67, 87.5, 100.0, 100.0, 100.0, 91.67, 91.67]))
+    assert_that(result['Unique Noun Count'][0: 10], equal_to([9, 2, 3, 3, 4, 10, 4, 2, 7, 5]))
+    assert_that(result_none_text['Lexical Density'], equal_to([np.nan]))
+    assert_that(result_none_text['Unique Noun Count'], equal_to([np.nan]))
```
Comment (Nadav-Barak) [inline tests/nlp/utils/test_properties.py:56]: you may need to add a line here for Code Style CI

**Review by Nadav-Barak — APPROVED** [review]

File: tests/nlp/utils/test_properties.py [hunk 9]
```diff
@@ -38,3 +37,20 @@ def test_calculate_default_properties():
 
     # Assert
     assert_that(result, equal_to({'Toxicity': [0] * 20_000}))
+
+
+def test_skipping_long_calculation_properties(tweet_emotion_train_test_textdata):
```
Comment (Nadav-Barak) [inline tests/nlp/utils/test_properties.py:42]: change the name of this test

File: deepchecks/nlp/utils/text_properties.py (hunk 5 above)
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done

File: deepchecks/nlp/utils/text_properties.py (hunk 6 above)
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done

File: deepchecks/nlp/utils/text_properties.py (hunk 7 above)
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done

File: tests/nlp/utils/test_properties.py (hunk 8 above)
Comment (hjain5164) [reply tests/nlp/utils/test_properties.py:69]: Done

File: tests/nlp/utils/test_properties.py (hunk 9 above)
Comment (hjain5164) [reply tests/nlp/utils/test_properties.py]: Done

File: deepchecks/nlp/utils/text_properties.py [hunk 10]
```diff
@@ -194,6 +196,39 @@ def formality(raw_text: Sequence[str], device: Optional[int] = None) -> List[flo
     return [x['score'] if x['label'] == 'formal' else 1 - x['score'] for x in classifier(raw_text)]
 
 
+def lexical_density(raw_text: Sequence[str]) -> List[str]:
+    """Return a list of floats of lexical density per text sample.
+
+    Lexical density is the percentage of unique words in a given text. For more
+    information: https://en.wikipedia.org/wiki/Lexical_density
+    """
+    nltk_download('punkt', quiet=True)
```
Comment (noamzbr) [inline deepchecks/nlp/utils/text_properties.py:205]: add an if statement here so that if download fails, you return raise a warning and return np.nan for all samples.
Same for next check

File: deepchecks/nlp/utils/text_properties.py [hunk 11]
```diff
@@ -251,7 +288,7 @@ def calculate_default_properties(
         The properties to calculate. If None, all default properties will be calculated. Cannot be used together
         with ignore_properties parameter. Available properties are:
         ['Text Length', 'Average Word Length', 'Max Word Length', '% Special Characters', 'Language',
-        'Sentiment', 'Subjectivity', 'Toxicity', 'Fluency', 'Formality']
+        'Sentiment', 'Subjectivity', 'Toxicity', 'Fluency', 'Formality', 'Lexical Density', 'Noun Count']
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:291]:
```suggestion
        'Sentiment', 'Subjectivity', 'Toxicity', 'Fluency', 'Formality', 'Lexical Density', 'Unique Noun Count']
```

File: docs/source/nlp/usage_guides/nlp_properties.rst [hunk 12]
```diff
@@ -57,6 +57,8 @@ Subjectivity                    Subjectivity of the text. Uses the textblob libr
 Toxicity*                       Toxicity of the text. Uses the unitary/toxic-bert model
 Fluency*                        Fluency of the text. Uses the prithivida/parrot_fluency_model model
 Formality*                      Formality of the text. Uses the s-nlp/roberta-base-formality-ranker model
+Lexical Density                 Percentage of unique words in the text, rounded up to 2 decimal digits
+Unique Nount Count              Number of unique noun words in the text
```
Comment (Nadav-Barak) [inline docs/source/nlp/usage_guides/nlp_properties.rst:61]:
```suggestion
Unique Noun Count*              Number of unique noun words in the text
```

**Review by Nadav-Barak — APPROVED** [review]

File: deepchecks/nlp/utils/text_properties.py (hunk 10 above)
Comment (hjain5164) [reply deepchecks/nlp/utils/text_properties.py]: Done

**Review by noamzbr — APPROVED** [review]

**Comment by noamzbr** [comment]
@all-contributors please add @hjain5164 for code
